src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp

```diff
@@ -391,13 +391,14 @@ class ShenandoahCardCluster: public CHeapObj<mtGC> {
   // a particular card region.  We pack this bit into start byte under assumption that start byte is accessed less
   // frequently than last byte.  This is true when number of clean cards is greater than number of dirty cards.
   static const uint16_t ObjectStartsInCardRegion = 0x80;
-  static const uint16_t FirstStartBits           = 0x3f;
+  static const uint16_t FirstStartBits           = 0x7f;
 
   crossing_info *object_starts;
 
 public:
   // If we're setting first_start, assume the card has an object.
   inline void set_first_start(size_t card_index, uint8_t value) {
+    assert(value < FirstStartBits, "Offset into card would be truncated");
```

Comment:
`value <= FirstStartBits` ?

Can/should replace (or augment, if you prefer) with a static assert using the maximum offset computed using card size and oop size?

Reply:
(dynamic) may be also in `set_last_start`?

Reply:
I like the idea of a static assert, will look into that.

Reply:
Shouldn't it be `<`? If the offset were `=` to `FirstStartBits` it should be offset `0` in the next card?